### DISCUSSION TITLE
Avoid sort block for MergeJoin when the join key is "sorted by" key

### DISCUSSION THREAD

**Discussion by Alex-Cheng** [ClickHouse/ClickHouse#33290] [Q&A] [answered]
When choose MergeJoin instead HashJoin by setting `join_algorithm='prefer_partial_merge'` and the join key is actually sort key (declared in create table SQL `create table ... order by <the key used for join operator>`) , I see the sortBlock() function is always invoked. As the column is sort key it is not necessary to sort it again.

![image](https://user-images.githubusercontent.com/1518453/147669830-9e3fe9ce-f288-4b47-9d9e-394415a68f79.png)

**Comment by alexey-milovidov** [comment, marked as answer]
This is true - if the data is already sorted, it is pointless to sort it again. But this optimization is not implemented. It still requires some effort to track the ordering properties of the streams of data and correctly join partially sorted streams.
